File: Project5/go/task3.go
/*
 * Impelement a word counter that can count the number of occurences of every word
 * in a text file in Go
 * 
 * Svita Kiran
 * 11/13/2025
 * CS333
*/

package main

import (
	"fmt"
	"strings"
	"bufio"
	"os"
	"regexp"
	"sort"
)

func main() {
	file, _ := os.Open(os.Args[1])
    defer file.Close()

    counts := make(map[string]int)
    re := regexp.MustCompile(`[^\w']+`)

    scanner := bufio.NewScanner(file)
    for scanner.Scan() {
        line := strings.ToLower(scanner.Text())
        line = re.ReplaceAllString(line, " ")
        for _, w := range strings.Fields(line) {
            counts[w]++
        }
    }

    type pair struct{ word string; count int }
    pairs := make([]pair, 0, len(counts))
    for w, c := range counts {
        pairs = append(pairs, pair{w, c})
    }

    sort.Slice(pairs, func(i, j int) bool { return pairs[i].count > pairs[j].count })

    for i := 0; i < len(pairs) && i < 20; i++ {
        fmt.Printf("%-10s %d\n", pairs[i].word, pairs[i].count)
    }
}

File: Project5/haskell/task3.hs
{- Impelementing a word counter that can count the number of occurences of every word
in a text file in Haskell

Svita Kiran
CS333
11/13/2025
-}

import System.Environment (getArgs)
import Data.Char (isAlpha, toLower)
import Data.List (sortOn)
import qualified Data.Map.Strict as Map

-- make all characters lowercase
normalize :: String -> String
normalize = map toLower . filter isAlpha

-- count words using Map
countWords :: [String] -> Map.Map String Int
countWords = foldr (\w -> Map.insertWith (+) w 1) Map.empty

-- print top frequency words
printTop :: Int -> Map.Map String Int -> IO ()
printTop n m = mapM_ printWord $ take n sorted
  where
    sorted = reverse $ sortOn snd $ Map.toList m
    printWord (w, c) = putStrLn $ w ++ "\t" ++ show c

main :: IO ()
main = do
    args <- getArgs
    case args of
        [filename] -> do
            content <- readFile filename
            let wordsList = map normalize $ words content
            let freqMap = countWords wordsList
            printTop 20 freqMap
        _ -> putStrLn "Usage: runhaskell WordCount.hs <filename>"